src/union.rs

```diff
@@ -10,106 +10,167 @@
+use petgraph::{algo, EdgeType};
+
 use pyo3::prelude::*;
 use pyo3::Python;
-use std::cmp::Ordering;
 
-/// [Graph] Return a new PyDiGraph by forming a union from`a` and `b` graphs.
-///
-/// The algorithm has three phases:
-///  - adds all nodes from `b` to `a`. operates in O(n), n being number of nodes in `b`.
-///  - merges nodes from `b` over `a` given that:
-///     - `merge_nodes` is `true`. operates in O(n^2), n being number of nodes in `b`.
-///     - respective node in`b` and `a` share the same weight
-///  - adds all edges from `b` to `a`.
-///     - `merge_edges` is `true`
-///     - respective edge in`b` and `a` share the same weight
-///
-/// with the same weight in graphs `a` and `b` and merged those nodes.
-///
-/// The nodes from graph `b` will replace nodes from `a`.
-///
-///  At this point, only `PyDiGraph` is supported.
-fn _digraph_union(
-    py: Python,
-    a: &PyDiGraph,
-    b: &PyDiGraph,
-    merge_nodes: bool,
-    merge_edges: bool,
-) -> PyResult<PyDiGraph> {
-    let first = &a.graph;
-    let second = &b.graph;
-    let mut combined = PyDiGraph {
-        graph: first.clone(),
-        cycle_state: algo::DfsSpace::default(),
-        check_cycle: false,
-        node_removed: false,
-        multigraph: true,
-    };
-    let mut node_map = HashMap::with_capacity(second.node_count());
-    let mut edge_map = HashSet::with_capacity(second.edge_count());
+type StablePyGraph<Ty> = StableGraph<PyObject, PyObject, Ty>;
 
-    let compare_weights = |a: &PyAny, b: &PyAny| -> PyResult<bool> {
-        let res = a.compare(b)?;
-        Ok(res == Ordering::Equal)
-    };
-
-    for node in second.node_indices() {
-        let node_index = combined.add_node(second[node].clone_ref(py))?;
-        node_map.insert(node.index(), node_index);
+fn find_node_by_weight<Ty: EdgeType>(
+    py: Python,
+    graph: &StablePyGraph<Ty>,
+    obj: &PyObject,
+) -> PyResult<Option<NodeIndex>> {
+    let mut index = None;
+    for node in graph.node_indices() {
+        let weight = graph.node_weight(node).unwrap();
+        if obj
+            .as_ref(py)
+            .rich_compare(weight, pyo3::basic::CompareOp::Eq)?
+            .is_true()?
+        {
+            index = Some(node);
+            break;
+        }
     }
+    Ok(index)
+}
 
-    for edge in b.weighted_edge_list(py).edges {
-        let source = edge.0;
-        let target = edge.1;
-        let edge_weight = edge.2;
-
-        let new_source = *node_map.get(&source).unwrap();
-        let new_target = *node_map.get(&target).unwrap();
+#[derive(Copy, Clone)]
+enum Entry<T> {
+    Merged(T),
+    Added(T),
+    None,
+}
 
-        let edge_index = combined.add_edge(
-            new_source,
-            new_target,
-            edge_weight.clone_ref(py),
-        )?;
+fn extract<T>(x: Entry<T>) -> T {
+    match x {
+        Entry::Merged(val) => val,
+        Entry::Added(val) => val,
+        Entry::None => panic!("called `Entry::extract()` on a `None` value"),
```

Comment:
Instead if doing this I wonder if it makes more sense to have a custom exception class with a user facing error. I realize it shouldn't ever happen at runtime because we iterate over the second graph and populate all the nodes in node_map so this is a pathological edge case. I'm just worried because a user would see ```PanicException: called `Entry::extract()` on a `None` value``` if this happened right now and it's not really user actionable or something they could handle. Maybe all we need to do is change the error message to be something like:

```suggestion
        Entry::None => panic!("Unexpected internal error: called `Entry::extract()` on a `None` value. Please file an issue at https://github.com/Qiskit/retworkx/issues/new/choose with the details on how you encountered this."),
```

Reply:
Well, if we ever encounter a `None` value here it will be a "serious" bug and probably from petgraph side, i'm totally ok to just panic here with the improved error message.